### PR TITLE
Send fleet-server elasticsearch config under new bootstrap attribute

### LINKED ISSUE #2784: Fleet Server configuration does not contain all the hosts available in the Elasticsearch cluster
- Version: 8.7.1 (and probably newer versions)
- Operating System: Linux
- Discuss Forum URL: N/A

Following the instructions to enroll a new Fleet Server, if the host used in the install command is down or stopped, fleet-server becomes unhealhty (and all the agents connected to it).

## Steps to Reproduce:

- Setup a Elastic stack cluster with 3 Elasticsearch hosts and 1 Kibana.
    - Ensure that the 3 ES hosts are accessible to port 9200
        - In this example, these ES hosts are accessible through:
            - https://10.13.233.1:9200
            - https://10.13.233.1:9201
            - https://10.13.233.1:9202
    - Ensure that Kibana has configured the 3 ES hosts in its configuration.
- Add in the Fleet Settings, all three hosts available in the Output
  fleet output configuration

- Create a new VM. In this VM, enroll a new fleet-server using the command suggested by the UI in another VM.
    - NOTE: there is just one ES host set in the `--fleet-server-es` parameter
    - Added some parameters to avoid issues with SSL/TLS
      ```bash
      sudo ./elastic-agent install \
        --fleet-server-es=https://10.13.233.1:9200 \
        --fleet-server-es-ca=/home/ubuntu/ca/ca.crt \
        --fleet-server-es-insecure \
        --insecure \
        --fleet-server-service-token=<token> \
        --fleet-server-policy=fleet-server-policy
      ```
- Create another VM. In this VM, enroll a new agent following the instructions too.
    - Adding `--insecure` to avoid issues with SSL/TLS
      ```
      sudo ./elastic-agent install \
        --url=https://fleetserver:8220 \
        --enrollment-token=<token> \
        --insecure
      ```
- Confirm that there are logs in Elasticsearch
  logs from fleet server and agent

At this point, this new fleet-server should be shown in the Fleet view along with the agents:
fleet healthy

In this stage, if any ES host different to the one used in the `install` command is stopped, there are no issues with fleet-server or any agent that has been installed:
- Agents and fleet server are healthy
- Agents and Fleet server are ingesting logs

However, if the ES host (`https://10.13.233.1:9200`) used in the install command (`--fleet-server-es`) is down:
- fleet-server becomes **unhealhty** as well as the agents.
- both fleet-server and agents are still sending data (logs) to Elasticsearch

unhealthy fleet server and agents

logs still are sent

Output of the `elastic-agent status`:
```
$ sudo elastic-agent status
State: HEALTHY
Message: Running
Fleet State: FAILED
Fleet Message: status code: 503, fleet-server returned an error: ServiceUnavailable, message: Fleet server unable to communicate with Elasticsearch
Components:
  * beat/metrics    (HEALTHY)
                    Healthy: communicating with pid '32799'
  * http/metrics    (HEALTHY)
                    Healthy: communicating with pid '32798'
  * system/metrics  (HEALTHY)
                    Healthy: communicating with pid '32761'
  * log             (HEALTHY)
                    Healthy: communicating with pid '32768'
  * fleet-server    (HEALTHY)
                    Healthy: communicating with pid '32714'
  * filestream      (HEALTHY)
                    Healthy: communicating with pid '32773'
```

Thanks to this PR https://github.com/elastic/fleet-server/pull/2632, when the fleet-server was healthy, a diagnostics bundle was generated and in the fleet-server configuration there is just one Elasticsearch host (the one used in the install command).

```yaml
fleet:
    agent:
        id: e1a89fa0-5c2a-41ed-a267-b615ebb34f92
        version: 8.7.1
        logging:
            level: info
    host:
        id: ""
        name: ""
output:
    elasticsearch:
        protocol: https
        hosts:
            - 10.13.233.1:9200
        path: ""
        headers: {}
        apikey: ""
        servicetoken: '[redacted]'
        proxyurl: ""
        proxydisable: false
        proxyheaders: {}
        tls:
            verification_mode: none
            certificate_authorities:
                - /home/ubuntu/ca/ca.crt
            renegotiation: never
        maxretries: 3
        maxconnperhost: 128
        timeout: 1m30s
        maxcontentlength: 104857600
...
```

### LINKED ISSUE #4769: Implement bootstrap test with fleet-server
The elastic-agent repo needs to have an automated test where it verifies the fleet-server bootstrap behaviour.

At a minimum this test should ensure that stand-alone installs function as expected.
Support for testing containers can be added at a later date

### PR DESCRIPTION
**Pull request by michel-laterman**
## What does this PR do?

Alter the fleet-server bootstrap component modifier to insert all (elasticsearch)
output configuration options specified by enrollment args under a new elasticsearch.boostrap key
instead of overwriting any existing keys. This will allow elastic-agent to send the list of hosts
(and other config options) retrieved from a policy to fleet server as well as the config needed
to form the initial connection to elasticsearch used to collect policy information.
Fleet-server will be altered to use the bootstrap config that is passed if the policy attributes
are unspecified or fail.

## Why is it important?

Fleet-server instances running under an elastic-agent are not able to connect to multiple ES hosts, or use any output settings that update local configuration that has been specified during enrolment. 

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- ~~I have made corresponding changes to the documentation~~
- ~~I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)
- [ ] I have added an integration test or an E2E test

## How to test this PR locally

See description for: https://github.com/elastic/fleet-server/pull/3506

## Related issues

- Closes #2784 
- Closes #4769 
- Relates https://github.com/elastic/fleet-server/issues/3464

### CODE REVIEW

**Comment by cmacknz** [comment]
Approach LGTM.

Do we have an automated test to ensure the agent can bootstrap a real fleet server already? If not we are going to need one.

**Comment by michel-laterman** [comment]
We don't have a full e2e test where we run a compiled agent with a fleet-server. This can be added in a followup PR as this change requires fleet-server's pr to also be merged.

We have partial integration tests in fleet-server where we fake agent communications to validate that the bootstrap sequence works as expected (i.e., state is degraded if no agent id is provided and goes healthy after): https://github.com/elastic/fleet-server/blob/main/internal/pkg/server/agent_integration_test.go

And fleet-server also has some disabled e2e tests (that never ran properly through ci) to test agent bootstrapping.

@blakerouse tells me that we should be able to add a fleet-server bootstrap test as part of the integration test suite in elastic-agent as long as we don't use containers.

**Comment by cmacknz** [comment]
We probably need a test that fleet server can be bootstrapped in both repositories, so we can prevent merging changes that would break it before they hit the snapshot image.

**Comment by michel-laterman** [comment]
Currently; when I try to use a binarry built with this PR in ESS I get
```
{"log.level":"error","@timestamp":"2024-05-09T16:47:16.527Z","log.origin":{"file.name":"coordinator/coordinator.go","file.line":608},"message":"Spawned new component fleet-server-default: input not supported","log":{"source":"elastic-agent"},"component":{"id":"fleet-server-default","state":"FAILED"},"ecs.version":"1.6.0"}
```

**Comment by cmacknz** [comment]
The spec file declaring the fleet-server input type is missing.

**Comment by michel-laterman** [comment]
OK, using a VM I was finally able to build and push an image for testing.
The image includes the corresponding fleet-server binary from https://github.com/elastic/fleet-server/pull/350

I have confirmed that:
- the agent running fleet-server starts and appears healthy in the UI.
- log levels can be changed on the ESS fleet-server instance, and the UI can be used to gather diagnostics bundles
- We are able to enroll a self-managed fleet-server instance, it also appears as healthy in the UI.
- a self-managed fleet-server instance can change log levels and collect diagnostics bundles
- the diagnostics bundle of a self-managed fleet-server can show multiple hosts in the output specification.

There are still some other automated tests we need to verify before merging these 2 PRs

File: changelog/fragments/1714505172-Send-fleet-server-component-output-under-bootstrap-key.yaml [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+# - breaking-change: a change to previously-documented behavior
+# - deprecation: functionality that is being removed in a later release
+# - bug-fix: fixes a problem in a previous version
+# - enhancement: extends functionality but does not break or fix existing behavior
+# - feature: new functionality
+# - known-issue: problems that we are aware of in a given version
+# - security: impacts on the security of a product or a user’s deployment.
+# - upgrade: important information for someone upgrading from a prior version
+# - other: does not fit into any of the other categories
+kind: feature
```
Comment (blakerouse) [inline changelog/fragments/1714505172-Send-fleet-server-component-output-under-bootstrap-key.yaml:11]:
```suggestion
kind: enhancement
```

File: changelog/fragments/1714505172-Send-fleet-server-component-output-under-bootstrap-key.yaml [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+# this field accommodate a description without length limits.
+# NOTE: This field will be rendered only for breaking-change and known-issue kinds at the moment.
+description: |
+  Alter the fleet-server bootstrap component modifier to insert all (elasticsearch)
+  output configuration options specified by enrollment args under a new elasticsearch.boostrap key
+  instead of overwriting any existing keys. This will allow elastic-agent to send the list of hosts
+  (and other config options) retrieved from a policy to fleet server as well as the config needed
+  to form the initial connection to elasticsearch used to collect policy information.
+  Fleet-server will be altered to use the bootstrap config that is passed if the policy attributes
+  are unspecified or fail.
```
Comment (blakerouse) [inline changelog/fragments/1714505172-Send-fleet-server-component-output-under-bootstrap-key.yaml:26]: Be less technical in this. This is what our users will read when wanting to upgrade, so tell them the benefits.

File: changelog/fragments/1714505172-Send-fleet-server-component-output-under-bootstrap-key.yaml [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+# - enhancement: extends functionality but does not break or fix existing behavior
+# - feature: new functionality
+# - known-issue: problems that we are aware of in a given version
+# - security: impacts on the security of a product or a user’s deployment.
+# - upgrade: important information for someone upgrading from a prior version
+# - other: does not fit into any of the other categories
+kind: feature
+
+# Change summary; a 80ish characters long description of the change.
+summary: Send fleet-server component output under bootstrap key
```
Comment (blakerouse) [inline changelog/fragments/1714505172-Send-fleet-server-component-output-under-bootstrap-key.yaml:14]: Make this less technical, be more user facing. Give a quick benefit description.

`Fleet Server component now uses policy output configuration to communicate with elasticsearch`

File: changelog/fragments/1714505172-Send-fleet-server-component-output-under-bootstrap-key.yaml [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+# PR URL; optional; the PR number that added the changeset.
+# If not present is automatically filled by the tooling finding the PR where this changelog fragment has been added.
+# NOTE: the tooling supports backports, so it's able to fill the original PR number instead of the backport PR number.
+# Please provide it if you are adding a fragment for a different PR.
+#pr: https://github.com/owner/repo/1234
+
+# Issue URL; optional; the GitHub issue related to this changeset (either closes or is part of).
+# If not present is automatically filled by the tooling with the issue linked to the PR number.
+issue: 2784
```
Comment (blakerouse) [inline changelog/fragments/1714505172-Send-fleet-server-component-output-under-bootstrap-key.yaml:39]: Should be the whole URL.

File: changelog/fragments/1714505172-Send-fleet-server-component-output-under-bootstrap-key.yaml [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+  are unspecified or fail.
+
+# Affected component; a word indicating the component this changeset affects.
+component:
+
+# PR URL; optional; the PR number that added the changeset.
+# If not present is automatically filled by the tooling finding the PR where this changelog fragment has been added.
+# NOTE: the tooling supports backports, so it's able to fill the original PR number instead of the backport PR number.
+# Please provide it if you are adding a fragment for a different PR.
+#pr: https://github.com/owner/repo/1234
```
Comment (blakerouse) [inline changelog/fragments/1714505172-Send-fleet-server-component-output-under-bootstrap-key.yaml:35]: Link this PR.

**Review by blakerouse — APPROVED** [review]
Simple change, like it!

Fix the changelog before merging.

**Review by pchila — APPROVED** [review]

**Comment by michel-laterman** [comment]
I've merged the [changes to fleet-server](https://github.com/elastic/fleet-server/pull/3506).
I'll merge this tomorrow to ensure the resulting artifacts/image is built with both changes as the agent will fail to bootstrap if the fleet-server changes are missing. Fleet-server will not fail if only those changes are present.

**Comment by michel-laterman** [comment]
Here are the issues to add bootstrap tests:
- https://github.com/elastic/elastic-agent/issues/4769
- https://github.com/elastic/fleet-server/issues/3552

**Comment by cmacknz** [comment]
I would like to see https://github.com/elastic/elastic-agent/issues/4769 done as part of this PR. You are going to have to do this test manually without it, may as well make the test framework do the work for you.

**Pull request merged**